File: src/main/java/com/jcloisterzone/feature/modifier/FamiliesModifier.java
package com.jcloisterzone.feature.modifier;

import com.jcloisterzone.game.setup.SetupQuery;

import io.vavr.collection.HashSet;
import io.vavr.collection.Set;

public class FamiliesModifier extends FeatureModifier<String> {

	public static final String CONFLICT = "";

	public FamiliesModifier(String selector, SetupQuery enabledBy) {
        super(selector, enabledBy);
    }

	@Override
    public String mergeValues(String a, String b) {
        if (a == null) return b;
        if (b == null) return a;
        if (a.equals("grey")) return b; // Gambler's Luck
        if (b.equals("grey")) return a; // Gambler's Luck
        if (a.equals("both") || b.equals("both")) return "both"; // Both Colors
    	return a.equals(b) ? a : CONFLICT;
    }

    @Override
    public String valueOf(String attr) {
        return attr;
    }
}

File: src/main/ts/com/jcloisterzone/feature/modifier/FamiliesModifier.ts
import type { SetupQuery } from "../../game/setup/SetupQuery.js";
import { FeatureModifier } from "./FeatureModifier.js";

/** Family-colour modifier (Family Feud; also written by Gambler's Luck dice).
 *  Merging: "grey" yields to the other colour, "both" wins, differing colours conflict. */
export class FamiliesModifier extends FeatureModifier<string> {
  static readonly CONFLICT = "";

  constructor(selector: string, enabledBy: SetupQuery | null) {
    super(selector, enabledBy);
  }

  mergeValues(a: string | null, b: string | null): string {
    if (a === null) return b as string;
    if (b === null) return a;
    if (a === "grey") return b; // Gambler's Luck
    if (b === "grey") return a; // Gambler's Luck
    if (a === "both" || b === "both") return "both"; // both colours
    return a === b ? a : FamiliesModifier.CONFLICT;
  }

  valueOf(attr: string): string {
    return attr;
  }
}
